### PR TITLE
Implementation of hoisted scripts

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

Adds support for `<script type="module" hoist>`

## Testing

Testing both dev and build.

## Docs

Yep

### CODE REVIEW

**Comment by FredKSchott** [comment]
- can you link to the RFC?
- can you give a description of what was implemented here? The discussion drifted a bit from the original issue so it would be good to understand in plain english what changes are being made here.

**Comment by matthewp** [comment]
@FredKSchott updated with more information.

File: examples/portfolio/src/pages/index.astro [hunk 1]
```diff
@@ -5,6 +5,8 @@ import Button from '../components/Button/index.jsx';
 import Nav from '../components/Nav/index.jsx';
 import Footer from '../components/Footer/index.jsx';
 import PorfolioPreview from '../components/PortfolioPreview/index.jsx';
+import Hello from '../components/Hello.astro';
+import Another from '../components/Another.astro';
```
Comment (matthewp) [inline examples/portfolio/src/pages/index.astro:9]: oops, I thought I had checked these out, note to myself to remove these changes.

File: examples/portfolio/src/pages/index.astro [hunk 2]
```diff
@@ -200,6 +202,10 @@ const featuredProject = projects[0];
         <h1 class="title">
           <small class="subtitle">The personal site of </small>Jeanine White
         </h1>
+        <div>
+          <Hello name="Wilbur" />
+          <Another />
```
Comment (matthewp) [inline examples/portfolio/src/pages/index.astro:207]: these too of course

**Comment by matthewp** [comment]
I added a link to the docs preview, relevant URL is here: https://astro-docs-git-script-hoist-lift-optimize-whatever-pikapkg.vercel.app/guides/scripts

File: packages/astro/src/compiler/transform/head.ts [hunk 3]
```diff
@@ -115,6 +115,116 @@ export default function (opts: TransformOptions): Transformer {
...
+                          }
+                        ]
+                      }
+                    ],
+                    start: 0,
+                    end: 0,
+                    children: [],
+                  },
+
+                  { // TODO change
```
Comment (tony-sull) [inline packages/astro/src/compiler/transform/head.ts:180]: The `data-astro` attribute setup here looks good to me, is this TODO comment still needed?

**Review by tony-sull — COMMENTED** [review]
Excellent, this looks good to me!  Just one small comment where I noticed a TODO comment hanging around

I'm not as familiar with the script bundling process during build, but both the codegen logic and test coverage look good so the build magic must work as well 😁

**Review by tony-sull — APPROVED** [review]
lgtm

File: docs/src/pages/guides/scripts.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
+---
+layout: ~/layouts/MainLayout.astro
+title: Scripts
+---
+
+Astro automatically adds script tags for your partially hydrated components and bundles those in production.
+
+Any other script that you want to run in your page you'll add it by adding a `<script>` somewhere in your page, or within a component. 
+
+# Hoisted scripts
```
Comment (FredKSchott) [inline docs/src/pages/guides/scripts.md:10]: I would recommend to move this section into `astro-components.md`, similar to how `### CSS Styles`  exists as a section in that doc.

File: packages/astro/src/@types/astro.ts [hunk 5]
```diff
@@ -67,12 +67,19 @@ export interface JsxItem {
...
+}
+
 export interface TransformResult {
   script: string;
   imports: string[];
   exports: string[];
+  components: string[];
   html: string;
   css?: string;
+  scripts: ScriptInfo[];
```
Comment (FredKSchott) [inline packages/astro/src/@types/astro.ts:82]: `script` and `scripts` feels confusing. `inlineScripts` since you have `hoistedScripts` below?

File: packages/astro/src/build/bundle/js.ts [hunk 6]
```diff
@@ -21,6 +24,159 @@ export function collectJSImports(buildState: BuildOutput): Set<string> {
...
+          if (imported) {
+            const outUrl = new URL(source, 'http://example.com' + imported);
+            return outUrl.pathname;
+          }
+
+          return null;
+        },
+        async load(id: string) {
+          if(virtualScripts.has(id)) {
+            return virtualScripts.get(id)!.content!;
```
Comment (FredKSchott) [inline packages/astro/src/build/bundle/js.ts:104]: these non-null assertion lint rules are worth fixing

File: packages/astro/src/build/bundle/js.ts [hunk 7]
```diff
@@ -21,6 +24,159 @@ export function collectJSImports(buildState: BuildOutput): Set<string> {
...
+    }
+  }
+
+  if(!entryImports.length) {
+    // There are no hoisted scripts, bail
+    return;
+  }
+
+  /**
+  * 2. Run the bundle to bundle each pages JS into a single bundle (with shared content)
```
Comment (FredKSchott) [inline packages/astro/src/build/bundle/js.ts:79]: There was an original idea of Astro that we were moving from "bundle JS by page" to "bundle JS by each hydrated component", using the hydrated components as "entrypoints"/"inputs" for your final bundled output.

This seems to be breaking that idea a bit, and moving into a page-based bundling world. Can we treat hoisted scripts the same way that we treat hydrated JS components, and just handle them in our current bundle step? It feels odd to be doing two different JS bundling passes (three if you count CSS bundling) when every other tool that I'm familiar with would just do a single pass.

**Review by FredKSchott — COMMENTED** [review]
Gave it a quick look and things generally look good! Only real question is around bundling complexity.

I'd also really like this to get @natemoo-re 's review given his work on the new compiler to support some compiler context / return value so that we don't need to encode the hoisted script into the final HTML like this. Would love to make sure you two are on the same page to make sure that transition is as easy as possible.

File: packages/astro/src/build/bundle/js.ts (hunk 7 above)
Comment (matthewp) [reply packages/astro/src/build/bundle/js.ts:79]: We could, but that's not how CSS works and was not the original ask in #370. What would be the advantage to having N script tags in this case?

One disadvantage to doing it the other way I see (aside from more requests) is that the current implementation results in hashed bundle names and the component bundling does not.

File: packages/astro/src/@types/astro.ts (hunk 5 above)
Comment (matthewp) [reply packages/astro/src/@types/astro.ts:82]: I'll change to `hoistedScripts` here too, inline wouldn't be right because they could be inline or external.

File: packages/astro/src/build/bundle/js.ts (hunk 7 above)
Comment (matthewp) [reply packages/astro/src/build/bundle/js.ts:79]: One advantage to doing the other way is consistency, just to be clear that I'm not completely opposed to it. I could see the argument for leaving it the other way (even though I don't think that's what users want and not how CSS works) and then proposing a change via RFC.

Comment (FredKSchott) [reply packages/astro/src/build/bundle/js.ts:79]: I don't think I understand enough about what this code is doing to answer well, will leave for you and @natemoo-re to decide the right path forward. Happy to jump in a convo with you both to discuss more, as well!

Comment (matthewp) [reply packages/astro/src/build/bundle/js.ts:79]: The reason for the 2 passes here is because the 2nd pass happens after the HTML has already been written and this bundling needs to update the HTML (to remove script tags and update to the hashed file names). In the future if we refactor the build, it should be to do all of the JS during this pass. Even though this is 2 passes, it's not recompiling the same code, these are different entry points.

File: docs/src/pages/guides/scripts.md (hunk 4 above)
Comment (matthewp) [reply docs/src/pages/guides/scripts.md]: Moved this
